From bce58809d87ff12c2a154633671399bcef355d8c Mon Sep 17 00:00:00 2001
From: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
Date: Mon, 17 Jul 2023 19:03:30 +0300
Subject: [PATCH 12/24] libxl/arm: Reuse generic PCI-IOMMU bindings for
 virtio-pci devices

Use the same "xen-grant-dma" device concept for the PCI devices
behind device-tree based PCI Host controller, but with one modification.
Unlike for platform devices, we cannot use generic IOMMU bindings
(iommus property), as we need to support more flexible configuration.
The problem is that PCI devices under the single PCI Host controller
may have the backends running in different Xen domains and thus have
different endpoints ID (backend domains ID).

Reuse generic PCI-IOMMU bindings (iommu-map/iommu-map-mask properties)
which allows us to describe relationship between PCI devices and
backend domains ID properly. Linux guest is already able to deal
with generic PCI-IOMMU bindings (see Linux drivers/xen/grant-dma-ops.c
for details).

According to Linux:
   - Documentation/devicetree/bindings/pci/pci-iommu.txt
   - Documentation/devicetree/bindings/iommu/xen,grant-dma.yaml

Signed-off-by: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
Upstream-Status: Inappropriate [downstream fork / BSP-layer delta; not submitted upstream]
Origin: https://github.com/xen-troops/xen commit bce58809d87ff12c2a154633671399bcef355d8c
---
 tools/libs/light/libxl_arm.c | 66 ++++++++++++++++++++++++++++++++++++
 1 file changed, 66 insertions(+)

diff --git a/tools/libs/light/libxl_arm.c b/tools/libs/light/libxl_arm.c
index ffd963736f..7fa3fa73a4 100644
--- a/tools/libs/light/libxl_arm.c
+++ b/tools/libs/light/libxl_arm.c
@@ -946,6 +946,7 @@ static int make_vpci_node_common(libxl__gc *gc, void *fdt,
 
 #define PCI_NUM_IRQ   4
 #define PCI_IRQ_MAP_STRIDE   8
+#define PCI_IOMMU_MAP_STRIDE 4
 
 static int create_vpci_irq_map(libxl__gc *gc, void *fdt)
 {
@@ -993,6 +994,67 @@ static int create_vpci_irq_map(libxl__gc *gc, void *fdt)
     return 0;
 }
 
+static int create_vpci_iommu_map(libxl__gc *gc, void *fdt,
+                                 libxl_domain_config *d_config)
+{
+    uint32_t *full_iommu_map = NULL;
+    uint32_t *iommu_map;
+    unsigned int i, ntranslated = 0;
+    int res;
+
+    for (i = 0; i < d_config->num_virtios; i++) {
+        libxl_device_virtio *virtio = &d_config->virtios[i];
+
+        if (libxl_defbool_val(virtio->grant_usage) &&
+            virtio->transport == LIBXL_VIRTIO_TRANSPORT_PCI) {
+            ntranslated++;
+        }
+    }
+
+    if (!ntranslated)
+        return 0;
+
+    full_iommu_map = malloc(ntranslated * sizeof(uint32_t) * PCI_IOMMU_MAP_STRIDE);
+    if (!full_iommu_map)
+        return -ENOMEM;
+
+    iommu_map = full_iommu_map;
+
+    /* See Linux Documentation/devicetree/bindings/pci/pci-iommu.txt */
+    for (i = 0; i < d_config->num_virtios; i++) {
+        libxl_device_virtio *virtio = &d_config->virtios[i];
+
+        if (libxl_defbool_val(virtio->grant_usage) &&
+            virtio->transport == LIBXL_VIRTIO_TRANSPORT_PCI) {
+            uint16_t bdf = (virtio->u.pci.bus << 8) |
+                (virtio->u.pci.dev << 3) | virtio->u.pci.func;
+            unsigned int j = 0;
+
+            /* rid_base (1 cell) */
+            iommu_map[j++] = cpu_to_fdt32(bdf);
+
+            /* iommu_phandle (1 cell) */
+            iommu_map[j++] = cpu_to_fdt32(GUEST_PHANDLE_IOMMU);
+
+            /* iommu_base (1 cell) */
+            iommu_map[j++] = cpu_to_fdt32(virtio->backend_domid);
+
+            /* length (1 cell) */
+            iommu_map[j++] = cpu_to_fdt32(1 << 3);
+
+            iommu_map += PCI_IOMMU_MAP_STRIDE;
+        }
+    }
+
+    res = fdt_property(fdt, "iommu-map", full_iommu_map,
+                       ntranslated * sizeof(uint32_t) * PCI_IOMMU_MAP_STRIDE);
+    if (res) return res;
+
+    free(full_iommu_map);
+
+    return 0;
+}
+
 static int make_vpci_node_virtio(libxl__gc *gc, void *fdt,
                                  libxl_domain_config *d_config)
 {
@@ -1006,6 +1068,10 @@ static int make_vpci_node_virtio(libxl__gc *gc, void *fdt,
     res = create_vpci_irq_map(gc, fdt);
     if (res) return res;
 
+    /* xen,grant-dma bindings */
+    res = create_vpci_iommu_map(gc, fdt, d_config);
+    if (res) return res;
+
     return 0;
 }
 
